src/trio/_core/_tests/test_cancel.py

```diff
@@ -0,0 +1,151 @@
+        task_status.started()
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to nursery with reason 'child task raised exception ValueError\(\)' from task {fail_task!r}$",
+        ):
+            await trio.sleep_forever()
+        raise TypeError
+
+    with RaisesGroup(ValueError, TypeError):
+        async with trio.open_nursery() as nursery:
+            fail_task = await nursery.start(failing_task)
+            await nursery.start(cancelled_task, fail_task)
+
+
+async def test_cancel_reason_nursery2() -> None:
+    async def failing_task(
+        event: trio.Event, task_status: trio.TaskStatus[trio.lowlevel.Task]
+    ) -> None:
+        task_status.started(current_task())
+        await event.wait()
+        raise ValueError
+
+    async def cancelled_task(
+        fail_task: trio.lowlevel.Task, task_status: trio.TaskStatus
+    ) -> None:
+        task_status.started()
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to nursery with reason 'child task raised exception ValueError\(\)' from task {fail_task!r}$",
+        ):
+            await trio.sleep_forever()
+        raise TypeError
+
+    with RaisesGroup(ValueError, TypeError):
+        async with trio.open_nursery() as nursery:
+            event = trio.Event()
+            fail_task = await nursery.start(failing_task, event)
+            await nursery.start(cancelled_task, fail_task)
+            event.set()
+
+
+async def test_cancel_reason_not_overwritten() -> None:
+    with trio.CancelScope() as cs:
+        cs.cancel()
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to explicit from task {current_task()!r}$",
+        ):
+            await trio.lowlevel.checkpoint()
+        cs.deadline = -inf
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to explicit from task {current_task()!r}$",
+        ):
+            await trio.lowlevel.checkpoint()
+
+
+async def test_cancel_reason_not_overwritten_2() -> None:
+    with trio.CancelScope() as cs:
+        cs.deadline = -inf
+        with pytest.raises(Cancelled, match=r"^cancelled due to deadline$"):
+            await trio.lowlevel.checkpoint()
+        cs.cancel()
+        with pytest.raises(Cancelled, match=r"^cancelled due to deadline$"):
+            await trio.lowlevel.checkpoint()
+
+
+async def test_nested_child_source() -> None:
+    ev = trio.Event()
+    parent_task = current_task()
+
+    async def child() -> None:
+        ev.set()
+        with pytest.raises(
+            Cancelled,
+            match=rf"^cancelled due to nursery with reason 'Code block inside nursery contextmanager raised exception ValueError\(\)' from task {parent_task!r}$",
+        ):
+            await trio.sleep_forever()
+
+    with RaisesGroup(ValueError):
+        async with trio.open_nursery() as nursery:
+            nursery.start_soon(child)
+            await ev.wait()
```

Comment:
By the way I think this and others could use `trio.testing.wait_all_tasks_blocked` instead (and avoid having an event).

Reply:
oooh yes, thank you!